source/Halibut/Util/AsyncEx/TaskExtensions.cs

```diff
@@ -0,0 +1,73 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Halibut.Util.AsyncEx
+{
+    internal static class TaskExtensions
+    {
+        /// <summary>
+        /// Waits for a) the task to complete, b) the timeout to pass or c) the cancellationToken to be triggered.
+        /// Wraps it all up so that we dont get UnobservedTaskExceptions 
+        /// </summary>
+        /// <param name="task">The task to wait on</param>
+        /// <param name="timeout">The amount of time to wait until we give up</param>
+        /// <param name="cancellationToken">Supports task cancellation</param>
+        /// <returns>The task if successful, otherwise a TimeoutException or OperationCanceledException</returns>
+        /// <exception cref="TimeoutException">If the timeout gets reached before the task completes</exception>
+        /// <exception cref="OperationCanceledException">The task was cancelled via the cancellation token</exception>
+        public static async Task TimeoutAfter(this Task task, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            var timeOutTask = Task.Delay(timeout);
```

Comment:
This can be a neat trick. I have seen it done before, but IIRC we had to change our implementation in a production system. We changed it to `using (var timeoutCancellationTokenSource = new CancellationTokenSource(timeout))` https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.-ctor?view=netframework-4.8#System_Threading_CancellationTokenSource__ctor_System_TimeSpan_

Instead of both a `Task.Delay(timeout)` and a `CancellationTokenSource`, both which consume unmanaged resources, you only have the `CancellationTokenSource` which you dispose with the `using` statement.